pkg/maps/lbmap/lbmap.go

```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
 	return updateService(fe, zeroValue)
 }
 
-// UpdateService adds or updates the given service in the bpf maps
-func UpdateService(fe ServiceKey, backends []ServiceValue, addRevNAT bool, revNATID int) error {
+// UpdateService adds or updates the given service in the bpf maps (in both -
+// legacy and v2).
+func UpdateService(fe ServiceKey, backends []ServiceValue,
+	addRevNAT bool, revNATID int,
+	acquireBackendID func(loadbalancer.L3n4Addr) (uint16, error),
+	releaseBackendID func(uint16)) error {
+
 	var (
 		weights         []uint16
 		nNonZeroWeights uint16
 		existingCount   int
+		backendKey      BackendKey
+		svcKeyV2        ServiceKeyV2
 	)
 
-	svc := cache.prepareUpdate(fe, backends)
+	// Acquire missing backend IDs
+
+	newBackendsByAddrID := map[BackendAddrID]ServiceValue{}
+	newBackendAddrIDs := map[BackendAddrID]struct{}{}
+	for _, b := range backends {
+		addrID := b.BackendAddrID()
+		newBackendAddrIDs[addrID] = struct{}{}
+		newBackendsByAddrID[addrID] = b
+	}
+	newBackendAddrIDs = cache.newBackendAddrIDs(newBackendAddrIDs)
+	newBackendIDs := map[BackendAddrID]uint16{}
+
+	for addrID := range newBackendAddrIDs {
+		addr := *serviceValue2L3n4Addr(newBackendsByAddrID[addrID])
+		backendID, err := acquireBackendID(addr)
+		if err != nil {
+			return fmt.Errorf("Unable to acquire backend ID for %s: %s", addrID, err)
```

Comment:
I was thinking about adding a code to release backend IDs in the case of an error, but in the end I didn't add it. Mainly because the local ID allocator returns the same ID for a backend which previously acquired and didn't release it. So, e.g. if a svc add request fails, a retry request will not acquire new IDs for the same backends.

Also in theory, `UpdateService` should never fail unless there is a bug in the code.

Reply:
Does this cause a resource management problem long-term? If you cycle backends enough times, then the ID allocator would eventually run out of IDs. It's keyed by IP address + port so difficult with IPv4 to run out of entries but plausible (10.x.y.z/8 = 24 bits + port = 16 bits + proto)

Reply:
(FWIW if we print a big angry warning message to highlight that this kind of issue then this is probably workable... What I don't want to do is to hit some really weird bug where LB stops working in some vague way far in the future and find it's impossible to trace down that this was the class of issue)

Reply:
With backend IDs of the u32 type it should be less of a problem.

Let me address this in a separate PR in which I'm making `UpdateService` idempotent.